src/PROPOSAL/detail/PROPOSAL/Propagator.cxx

```diff
@@ -139,93 +142,112 @@ Interaction::Loss Propagator::DoStochasticInteraction(ParticleState& p_cond,
-    std::tie(mean_direction, new_direction) = utility.DirectionsScatter(
-        grammage_next_interaction, state.energy, E_f, state.direction, rnd);
-
-    double distance_next_interaction;
-    try {
-        distance_next_interaction = density->Correct(state.position,
-            mean_direction, grammage_next_interaction, max_distance_left);
-    } catch (const DensityException&) {
-        Logging::Get("proposal.propagator")
-            ->debug("Interaction point exceeds "
-                    "maximum propagation distance or lies in infinity.");
-        distance_next_interaction = INF;
-    }
-    auto new_position
-        = state.position + distance_next_interaction * mean_direction;
-
-    auto AdvanceDistance = std::array<double, 3> { 0., 0., 0. };
-    AdvanceDistance[ReachedInteraction] = distance_next_interaction;
-    AdvanceDistance[ReachedMaxDistance] = max_distance_left;
-    AdvanceDistance[ReachedBorder]
-        = CalculateDistanceToBorder(state.position, mean_direction, *geometry);
-
-    int advancement_type = minimize(AdvanceDistance);
-    double advance_distance = AdvanceDistance[advancement_type];
-    double advance_grammage = grammage_next_interaction;
-
-    if (advancement_type != ReachedInteraction) {
-        double control_distance;
-        double energy_next_interaction = E_f;
-        do {
-            advance_distance = AdvanceDistance[advancement_type];
-            if (advancement_type == ReachedInteraction) {
-                // avoid recalculation if we go back to this case
-                advance_grammage = grammage_next_interaction;
-                E_f = energy_next_interaction;
-            } else {
-                advance_grammage = density->Calculate(
-                        state.position, state.direction, advance_distance);
-                E_f = utility.EnergyDistance(state.energy, advance_grammage);
-            }
+    // Calculate maximal allowed length of step (limit due to final_distance)
+    const double max_distance = final_distance - state.propagated_distance;
 
-            std::tie(mean_direction, new_direction) = utility.DirectionsScatter(
-                advance_grammage, state.energy, E_f, state.direction, rnd);
+    // Calculate grammage until next stochastic interaction
+    double grammage_next_interaction = utility.LengthContinuous(
+            state.energy, energy_next_interaction);
 
+    int advancement_type;
+    Cartesian3D mean_direction, new_direction; // proposed scattering
+
+    // This lambda expression ensures we always use the same 4 random numbers
+    std::array<double, 4> random_numbers = {-1, -1, -1, -1};
+    int i = 0;
+    auto rnd = [&rnd_generator, &i, &random_numbers]() {
+        if (random_numbers[i%4] == -1)
+            random_numbers[i%4] = rnd_generator();
+        return random_numbers[i++%4];
+    };
+
+    do {
+        // Calculate grammage, energy and distance for step
+        if (energy != -1 && grammage == -1 && distance == -1) {
```

Comment:
Code relict from when it was not clear what cases are actually existing.
The -1 indicates that a variable is not known. But since this is always the case for grammage, we can delete it.